File: go/practice16/go/CocktailSort.go
package main

import (
	"encoding/csv"
	"fmt"
	"os"
	"strconv"
	"time"
)

func main() {
	sizes := []int{100, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 20000, 30000, 40000, 50000}
	times := [][]float64{}

	for k := 0; k < 5; k++ {
		times = append(times, []float64{})
		fmt.Println(":::::::::", k)
		for i := 0; i < len(sizes); i++ {
			list := []int{}
			fileName := "data" + strconv.Itoa(sizes[i]) + ".csv"

			list = Read(fileName)

			start := time.Now()
			cocktailSort(list)

			times[k] = append(times[k], time.Since(start).Seconds())

			fmt.Println("finished: ", sizes[i])
		}
	}
	times = append(times, []float64{})
	Write(times)
}

func Read(filePath string) []int {
	list := []int{}
	file, err := os.Open(filePath)

	if err != nil {
		fmt.Println("Unable to read input file "+filePath, err)
	}

	defer file.Close()
	csvReader := csv.NewReader(file)
	data, err := csvReader.ReadAll()

	if err != nil {
		fmt.Println("Unable to parse file as CSV for "+filePath, err)
	}

	for i := 0; i < len(data); i++ {
		for j := 0; j < len(data[i]); j++ {
			number, err := strconv.Atoi(data[i][j])
			if err == nil {
				list = append(list, number)
			}
		}
	}

	return list
}

func Write(data [][]float64) {
	records := [][]string{{}, {}}

	file, err := os.Create("cocktailSortTimesGo.csv")
	defer file.Close()

	if err != nil {
		fmt.Println("failed to open file", err)
	}

	writer := csv.NewWriter(file)
	defer writer.Flush()

	for i := 0; i < len(data[0]); i++ {
		j := 0.0
		for k := 0; k < 5; k++ {
			j += data[k][i]
		}
		data[5] = append(data[5], j/5.0)
	}

	for i := 0; i < 6; i++ {
		records = append(records, []string{})
		for k := 0; k < len(data[i]); k++ {
			j := data[i][k]
			records[i] = append(records[i], strconv.FormatFloat(j, 'E', -1, 64))
		}
	}

	for _, record := range records {
		if err := writer.Write(record); err != nil {
			fmt.Println("error writing record to file", err)
		}
	}
}

func cocktailSort(list []int) {
	last := len(list) - 1
	for {
		swapped := false
		for i := 0; i < last; i++ {
			if list[i] > list[i+1] {
				list[i], list[i+1] = list[i+1], list[i]
				swapped = true
			}
		}
		if !swapped {
			return
		}
		swapped = false
		for i := last - 1; i >= 0; i-- {
			if list[i] > list[i+1] {
				list[i], list[i+1] = list[i+1], list[i]
				swapped = true
			}
		}
		if !swapped {
			return
		}
	}
}

File: go/practice16/cpp/CocktailSort.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <ctime> 
#include <sstream>
#include <fstream>
using namespace std;

void PrintVector(vector<int> list) {
  cout<<"[ ";
  for(int i = 0; i < list.size(); i++) cout<<list[i]<<" ,";
  cout<<"]"<<endl;
}

vector<int> cocktailSort(vector<int> list) {
	int last = list.size() - 1;
  while(true) {
		bool swapped = false;
		for (int i = 0; i < last; i++) {
			if (list[i] > list[i+1]) {
        int a =	list[i];
        list[i] = list[i+1];
        list[i+1] = a;
				swapped = true;
			}
		}
		if(!swapped) return list;
		
		swapped = false;
		for (int i = last - 1; i >= 0; i--) {
			if (list[i] > list[i+1]) {
				int a =	list[i];
        list[i] = list[i+1];
        list[i+1] = a;
				swapped = true;
			}
		}
		if (!swapped) return list;
  }
  return list;
}


vector<int> Read(string fileName) {
  vector<int> list;
  vector<vector<string>> content;
  vector<string> row;
  string line, word;
  fstream file(fileName, ios::in);
  if(file.is_open()) {
    while(getline(file, line)) {
      row.clear();
      stringstream str(line);
      while(getline(str, word, ',')) row.push_back(word);
      content.push_back(row);
    }
  }
  else {
    cout<<"ERROR!!"<<endl;
    return list;
  }
  
  for(int i = 0; i < content.size(); i++) {
    for(int j = 0; j < content[i].size(); j++) {
      list.push_back(stoi(content[i][j]));
    }
  }
  return list;
}


int main() {
  vector<int> sizes = { 100, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 20000, 30000, 40000, 50000 };
  vector<vector<double>> times;
  ofstream myFile("cocktailSortTimesCpp.csv");

  for(int k = 0; k < 5; k++) {
    cout<<":::::::::"<<k<<endl;
    vector<double> subTime;
    times.push_back(subTime);
    for(int i = 0; i < sizes.size(); i++) {
      vector<int> list = Read("data" + to_string(sizes[i]) + ".csv");
      auto startTime = chrono::system_clock::now();
      list = cocktailSort(list);
      auto endTime = chrono::system_clock::now();
      std::chrono::duration<double> elapsed_seconds = endTime - startTime;
      times[k].push_back(elapsed_seconds.count());

      cout<<"finished: "<< sizes[i]<<endl;
    }

    string cad = "";
    
    for(int i = 0; i < sizes.size(); i++) cad += to_string(times[k][i])+",";
    myFile << cad + "\n";
  }

  string cad = "";
  vector<double> subTime;

  times.push_back(subTime);
  for(int i = 0; i < times.size(); i++) {
    double avg = 0.0;
    for(int j = 0; j < 5; j++) avg += times[j][i];
    times[5].push_back(avg/5.0);
  }

  for(int i = 0; i < sizes.size(); i++) cad += to_string(times[5][i])+",";
  myFile << cad;
  myFile.close();

  return 0;
}
